File: opentt/src/main/java/com/bytedance/sdk/open/aweme/utils/Md5Utils.java
package com.bytedance.sdk.open.aweme.utils;

import java.security.MessageDigest;

/**
 * MD5 工具类
 * Created by yangzhirong on 2018/10/8.
 */
public class Md5Utils {

    private static final char[] hexDigits = new char[]{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};

    public Md5Utils() {
    }

    public static String hexDigest(String string) {
        String res = null;
        try {
            res = hexDigest(string.getBytes());
        } catch (Exception e) {
            e.printStackTrace();
        }
        return res;
    }

    public static String hexDigest(byte[] bytes) {
        String res = null;

        try {
            MessageDigest messageDigest = MessageDigest.getInstance("MD5");
            messageDigest.update(bytes);
            byte[] tmp = messageDigest.digest();
            char[] charStr = new char[32];
            int k = 0;
            for (int i = 0; i < 16; ++i) {
                byte b = tmp[i];
                charStr[k++] = hexDigits[b >>> 4 & 15];
                charStr[k++] = hexDigits[b & 15];
            }
            res = new String(charStr);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return res;
    }
}

File: openoversea/src/main/java/com/bytedance/sdk/open/tiktok/utils/Md5Utils.kt
package com.bytedance.sdk.open.tiktok.utils

/*
    Copyright 2022 TikTok Pte. Ltd.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */

import java.security.MessageDigest

internal object Md5Utils {
    private val hexDigits = charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f')

    fun hexDigest(string: String): String? {
        var res: String? = null
        try {
            res = hexDigest(string.toByteArray())
        } catch (e: Exception) {
            e.printStackTrace() // TODO: chen.wu remove?
        }
        return res
    }

    fun hexDigest(bytes: ByteArray): String? {
        var res: String? = null
        try {
            val messageDigest = MessageDigest.getInstance("MD5")
            messageDigest.update(bytes)
            val tmp: ByteArray = messageDigest.digest()
            val charStr = CharArray(32)
            var k = 0
            for (i in hexDigits.indices) {
                val b = tmp[i].toInt()
                charStr[k++] = hexDigits[(b ushr 4) and 15]
                charStr[k++] = hexDigits[(b and 15)]
            }
            res = String(charStr)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return res
    }
}
